### ISSUE TITLE
Little smarter getting IPv6 address

### ISSUE THREAD

**Issue by chris001** [virtualmin/virtualmin-gpl#389]
I see Virtualmin picked up a link-local IPv6 address (starting with `FE8`) from the machine's interface, and decided to make it the default IPv6 address to assign to any and all virtual servers.  This is a bug, a server using this IPv6 address can't and won't pass traffic to/from the internet over IPv6 yet Virtualmin thinks it can, which is not right.

I'm going to paste this here, it's a classification of 7 types of IPv6 address, some of which the code should probably get a little smarter about and never pick them as default IPv6 addresses for virtual servers.

1. `::1/128`  `localhost` aka loopback address, same as `127.0.0.1/8` or `127.x.x.x` in IPv4.

2. `fe80::/10` Link Local Address, same as `169.254.0.0/16`. Routers do not forward packets.

3. `2000::/3`.  Global unicast, same as IPv4 public IP address. Allocated by IANA. 

4. `::/128` is same as `0.0.0.0` in IPv4.
 Used as source address when a unique address has not yet been determined.
 Never assigned to interface or used as destination address.
 In terms of routing, `::/128` indicates default route, same as `0.0.0.0`.
with subnet mask `0.0.0.0` in IPv4.

5.  Site-local are link private IP4 address.
  Always begin with `FEC0`
  Replaced with ULA, which is `FC00::/7` (below).

6. IPv6 Unique Local Address block (ULA)
 Format is `FC00::/7`  Same type of local address as `10.0.0.0/8` and `192.168.0.0/16` in IPv4.
 Used inside private site or org.
 Routable only within the private network, will not get out on to the internet.

7. IPv4 compatible
  Special address assigned to devices that handle both IPv4 and IPv6 traffic that use one type of address, have it be an IPv6 address, so all the IPv4 traffic is actually going over IPv6.
  Start with 96-bit zeroes, followed by 32 bit IPv4 address.
  If your IPv4 is `123.25.135.1` then your IPv6 would be `::123.25.135.1` but you have to convert that to hex which is  `::7B19:8701` because it needs to be hex for IPv6.

Don't default to types 1, 2, 4, 5, or 6.  Types 3 and 7 should work fine.

**Comment by jcameron** [comment]
Thanks for the suggestions! I will implement this for the next release of Virtualmin.

Is a "Global unicast" address safe to use as the default v6 though?

**Comment by chris001** [comment]
> Is a "Global unicast" address safe to use as the default v6 though?

Yes, it's safe to use as the default IPv6 because "IPv6 Global Unicast addresses" are similar to IPv4 public IP addresses.  As the name implies, they are routable on the internet.  Currently IANA has assigned only `2000::/3` addresses to the global pool.  They may assign other prefixes in the future?

A global unicast IPv6 address consists of two parts:

    subnet ID – 64 bits long. Contains the site prefix (obtained from a Regional Internet Registry) and the subnet ID (subnets within the site).
    interface ID – 64 bits long. typically composed of a part of the MAC address of the interface.

Here is a graphical representation of the two parts of an global IPv6 address:
![global_unicast_ipv6_address](https://user-images.githubusercontent.com/259416/166468664-81945096-923a-400a-aae3-380885d653f6.jpg)

**Comment by jcameron** [comment]
You can change the detected default IPv6 address at System Settings -> Virtualmin Configuration.

**Comment by iliaross** [comment]
Reopened, as this was closed by an automated cleanup run.

⚠️ If the issue is resolved, please take a moment to close it to help keep the project organized.
